### DISCUSSION TITLE
Strange CPU load during night

### DISCUSSION THREAD

**Discussion by djtremolo** [blakeblackshear/frigate#7790] [Ask A Question] [answered]
This is not a road blocker, I just would want to understand the reason.

My Frigate setup now works very stable and well. I have Reolink 520A cameras on their maximum picture quality: 2560x1920 at 30FPS. That works fine.

However, when sun sets and the cameras go into low light mode, they drop the FPS automatically (not possible to configure) to ~10FPS and the stream goes greyscale.

My thinking is, now with greyscale 10FPS, the Frigate should consume *less* CPU. But it does not - actually the CPU load jumps significantly upward. This happens every night and has always been like that.

Now to the question:
- what would explain this behaviour?
- is there something that I can do to avoid it?

![Screenshot_20230913_210001_Home Assistant.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/c54da53c-d8a1-4af0-ab91-f587e0fae303)

ps. my server has enough cores and the night time usage still stays below maximum. The graph is from home assistant - that has no effect on the result.

**Comment by NickM-27** [comment]
it would be better to see the systme page at this point, so we can see which process is actually using more / less CPU.

it is a shame reolink doesn't allow customization of that on their 520A, I disabled that on my reolink doorbell because it just creates choppy bad video at night.

**Reply by djtremolo** [reply]
Sure! I skipped most of the cameras, they are almost identical.

I'll post a similar set tomorrow at daytime for comparision.

![Screenshot_20230913_212218_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/c585b3cd-cecb-4c5f-ba20-7eca4d3abb26)

![Screenshot_20230913_212307_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/0430134d-ba5b-4d55-a084-9a13437a227c)

![Screenshot_20230913_212239_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/ceb0ba60-e239-4b24-92d0-9214634c3b8b)

**Reply by NickM-27** [reply]
Well there you go

> My thinking is, now with greyscale 10FPS, the Frigate should consume less CPU. But it does not - actually the CPU load jumps significantly upward. This happens every night and has always been like that.

you were assuming the CPU was due to decoding but actually the increased cpu usage seems to be because motion detection CPU usage has jumped up. I also notice the go2rtc CPU usage is very high which is likely part of the issue as well. Would be good to know how both of those compare to during the day

**Reply by djtremolo** [reply]
Good catch! That's certainly true. Naturally, the low light photos will have more random noise due to high ISO and therefore, the detection will be busier 🤔

I wonder if this change in detection load is because of 
- motion detection (done in software, I suppose?) 
or 
- object detection (cuda)

And if by motion detection, would some parameter in frigate yaml have an effect on this?

Daytime screenshots coming up tomorrow 👍

**Reply by NickM-27** [reply, marked as answer]
> object detection (cuda)

no, because as you can see in the page the detect fps is 0

> And if by motion detection, would some parameter in frigate yaml have an effect on this?

you could maybe try turning improve contrast off at night. But in any case frigate motion detection has been rewritten for the next release

**Reply by djtremolo** [reply]
... about go2rtc. I'm afraid this is heavily offtopic, but just for context:

In my setup, all camera streams are taken into go2rtc first and then restreamed to multiple users. That's a long story including an external zoneminder etc, so here's the brief version:

For most of the cameras, I fetch only the main stream to the Frigate server and recode it to sub streams locally (by help of go2rtc) because then I can have better quality 1280x720/8fps for detection, whereas the camera sub stream would be 640x480. Also, the reolink cameras are more stable when they can work for only one stream.

My GPU would have the encoding power for doing all of that recoding, but the GPU memory becomes the bottleneck. Thus I can't run all cameras through NVidia and have to use i7 VAAPI for 2-3 camera sub streams. That works.

Overall, Go2rtc runs happily even though it is super busy. Most of the work is still offloaded to hw, so *during daytime* all of the CPU cores stay somewhere at <60% load.

**Reply by djtremolo** [reply]
*...you could maybe try turning improve contrast off at night. But in any case frigate motion detection has been rewritten for the next release*

Awesome! Looking forward into trying the next version! In the meanwhile, I entertain myself by tuning up the contrast setting.

Thank you again for the great support 👍

**Reply by djtremolo** [reply]
Here are the screenshots from today:

![Screenshot_20230914_082925_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/df9aacc4-1a42-4e94-af91-8077354a4fed)

![Screenshot_20230914_082943_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/3a17d4c9-83d6-44b7-ba44-31a73431e9f6)

![Screenshot_20230914_082956_Chrome.jpg](https://github.com/blakeblackshear/frigate/assets/23290782/e9915d3b-f26e-46ef-8a54-43f1ebee3095)

**Reply by NickM-27** [reply]
yeah, looks to just be a difference in motion detection

**Reply by djtremolo** [reply]
Still getting back to this subject. This is not Frigate specific anymore, more like "look what I found!" kind of thing:

In the picture below, you can see the Container CPU usage. The CPU has 6 cores = 12 threads, so the max load is 1200%. The ~600% therefore is somewhat half of the CPU capacity, which is fine.

However, see what happens in the core temperatures? When the load goes up during the night, the core temperatures decrease. And when the load goes down for daytime, the core temperatures get higher. Strange, huh? The ambient temperature stays pretty much in the range of 15-17'c for the whole time (running in my temp controlled garage).

Note: the strange spike in the graphs became because of a server reboot
Note: the data is collected by Glances at the server and visualized by Home Assistant.

![image](https://github.com/blakeblackshear/frigate/assets/23290782/da590fe5-409b-4050-a36a-67bf86e75cbb)

**Reply by mesistostefre** [reply]
well, likely the fan gets triggered cooling the CPUs a bit more down ...

**Reply by djtremolo** [reply]
Good point. That probably explains it 👍

**Comment by mesistostefre** [comment]
Sorry for jumping on this ... I noticed a similar behaviour with the 820A as soon as it switches to IR. 

Is it possible that the significantly increased noise in the IR stream makes Frigate think there is loads of motion? Which will make it send over to detection?

@djtremolo : I do not see the cam reduce its frame rate when in IR mode. Did not find that info on reolink's web pages. Have a link to doc that confirms this?

The other strange thing is: as soon as the cam switches to IR, Frigate receives and displays the stream even though the camera has been disabled: "enabled=false". But that is maybe a different subject and therefor off topic here...

**Reply by NickM-27** [reply]
> Is it possible that the significantly increased noise in the IR stream makes Frigate think there is loads of motion? Which will make it send over to detection?

like I said above to the OP, the detect fps in the screenshot was 0 which means that object detection is not being run

> @djtremolo : I do not see the cam reduce its frame rate when in IR mode. Did not find that info on reolink's web pages. Have a link to doc that confirms this?

this is what they mean with their "frame rate mode" being auto or constant. auto reduces frame rate in night mode while constant keeps it constant the entire time
